tests/test_codec.py

```diff
@@ -0,0 +1,170 @@
+        self.assertEqual(s8.encode(), 5)
+
+    def test_encode_canmatrix(self):
+        db_path = os.path.join(
+            os.path.dirname(__file__), "..", "test", "test.dbc")
+        for bus in formats.loadp(db_path).values():
+            test_frame1 = 0x123
+            data = {
+                'Signal': 2,
+                'someTestSignal': 101,
+            }
+            data_bytes = tuple(bytearray(bus.encode(test_frame1, data)))
+            assert data_bytes == (0, 0x28, 0x04, 0, 0, 0, 0, 0)
+
+    def test_decode_signal(self):
+        s1 = Signal('signal', signalSize=8)
+        self.assertEqual(s1.decode(1), 1)
+        self.assertEqual(s1.decode(127), int(s1.max))
+        self.assertEqual(s1.decode(-128), int(s1.min))
+
+        s2 = Signal('signal', signalSize=10, is_signed=False)
+        self.assertEqual(s2.decode(10), 10)
+        self.assertEqual(s2.decode(s2.max), 1023)
+        self.assertEqual(s2.decode(s2.min), 0)
+
+        s3 = Signal('signal', signalSize=8, factor=2)
+        self.assertEqual(s3.decode(5), 10)
+        self.assertEqual(s3.decode(127), int(s3.max))
+        self.assertEqual(s3.decode(-128), int(s3.min))
+
+        s4 = Signal('signal', signalSize=8, is_signed=False, factor=5)
+        self.assertEqual(s4.decode(2), 10)
+        self.assertEqual(s4.decode(255), int(s4.max))
+        self.assertEqual(s4.decode(0), int(s4.min))
+
+        s5 = Signal('signal', signalSize=8, offset=2)
+        self.assertEqual(s5.decode(8), 10)
+        self.assertEqual(s5.decode(127), int(s5.max))
+        self.assertEqual(s5.decode(-128), int(s5.min))
+
+        s6 = Signal('signal', signalSize=8, is_signed=False, offset=5)
+        self.assertEqual(s6.decode(5), 10)
+        self.assertEqual(s6.decode(255), int(s6.max))
+        self.assertEqual(s6.decode(0), int(s6.min))
+
+    def test_encode_decode_signal_value(self):
+        db_path = os.path.join(
+            os.path.dirname(__file__), "..", "test", "test.dbc")
+        for bus in formats.loadp(db_path).values():
+            test_frame1 = 0x123
+
+            data = {
+                'Signal': 2,
+                'someTestSignal': 101,
+            }
+            data_bytes = tuple(bytearray(bus.encode(test_frame1, data)))
+            decoded = bus.decode(test_frame1, data_bytes)
+
+            for k, v in data.items():
+                assert decoded[k] == v
+
+    def test_encode_decode_signal_value_choice_unicode(self):
+        db_path = os.path.join(
+            os.path.dirname(__file__), "..", "test", "test.dbc")
+        for bus in formats.loadp(db_path).values():
+            test_frame1 = 0x123
+
+            data = {
+                'Signal': u'two'
+            }
+            data_bytes = tuple(bytearray(bus.encode(test_frame1, data)))
+
+            decoded = bus.decode(test_frame1, data_bytes)
+
+            for k, v in data.items():
+                assert str(decoded[k]) == v
+
+    def test_encode_decode_signal_value_choice_str(self):
```

Comment:
In py2, `str` is a byte sequence and, basically, should not be used.  It does wacky automatic stuff on top of just being confusing because of it's misleading name.

We probably want a single function and:
```
from __future__ import unicode_literals
from builtins import *
```
http://python-future.org/compatible_idioms.html#unicode-text-string-literals
http://python-future.org/compatible_idioms.html#basestring

If someone wants to encode a string to bytes, they need to pick their encoding which we can't do for them usefully.

I don't generally like the `*` in `from builtins import *` but I sort of feel like this is a special case where we want to make certain that as much as possible is made to be common.

Reply:
This is not encoding a string to bytes. `'two'` is a value description for the signal `Signal` in the frame `0x123`.

The value getting encoded will be `2`

Reply:
Yeah, pretty sure I misread this too...  again, I'm sorry.

Still, I think the imports are relevant for consistency of the environment the code runs in.  `str` should be a Unicode character sequence in keeping with elsewhere.  I may have messed up not importing `unicode_literals` elsewhere, I'm not sure.

Also, I don't think we want to be testing to make sure that people can pass byte sequences (py2 `str`) where we expect them to pass a Unicode sequence.  Sure, py2 does goofy stuff but it shouldn't work in py3.  Presently I'm thinking this would pass in py2 because of the automatic bytes->unicode conversion and in py3 because of `''` and `u''` being the same thing anyways.  With the imports I would expect both to pass in both py2 and py3 because they are identical at that point.  If you changed one to be `b''` then it would fail in both py2 and py3.  If I'm predicting correctly...

Reply:
I fully get your point now. SO you suggest removing `test_encode_decode_signal_value_choice_str`?

Reply:
Along with adding the imports, sure.